### ISSUE TITLE
Moving project to Gradle

### ISSUE THREAD

**Issue by leventov** [apache/incubator-druid#7866]
There is an idea to move the project to Gradle, for two reasons:
 1) Maven builds take too long to complete. Gradle's incremental builds are much more powerful in recent years at cutting off the unnecessary work.
 2) We can run IntelliJ inspections locally using a plugin: https://github.com/JetBrains/inspection-plugin

This message was originally posted here: https://github.com/apache/incubator-druid/issues/7834#issuecomment-500963555

Note: I don't plan to work on this in the near future.

**Comment by drcrallen** [comment]
composite build in gradle is quite nice, and would allow people to setup a mono repo with druid as one of the components.

It *might* be possible to have a gradle build but as a Maven plugin rather than pure gradle. Would that work in this case?

**Comment by ccaominh** [comment]
I've started looking into this and initial results show that gradle 5.5 is about 3.5-4x faster than maven 3.6.1 for a clean build:

Non-parallel build:
- `mvn install -DskipTests -Dcheckstyle.skip` (4m 44s): https://scans.gradle.com/s/4sen23iys23sg
- `gradlew assemble publishToMavenLocal` (1m 20s / **3.5x speedup**): https://scans.gradle.com/s/oxuy353qrxkhw

With parallel builds enabled on an 8 core machine:
- `mvn install -DskipTests -Dcheckstyle.skip -T8` (2m 29s): https://scans.gradle.com/s/i3pkohokc4mtw
- `gradlew assemble publishToMavenLocal --parallel` (36s / **4.1x speedup**): https://scans.gradle.com/s/oo3qlngwnmrr4

Dependencies were downloaded before running measurements, and the gradle build cache was not enabled.

**Comment by himanshug** [comment]
@ccaominh that sounds very encouraging. I am not familiar with Gradle as much, do you know what gradle does smartly to make "clean" builds faster than maven ?

I am guessing that you disabled tests and checkstyle because they take alot of time and difference doesn't remain as noticeable with those being enabled.

was doing a bit of searching around and hit https://phauer.com/2018/moving-back-from-gradle-to-maven/  which has some valid points to consider as well.

**Comment by ccaominh** [comment]
@himanshug Gradle has a good article that describes some of the things they do to be faster than maven: https://gradle.org/gradle-vs-maven-performance/#performance-advantages-over-maven-that-make-this-possible

Regarding the tests, I'm still working on fixing some build issues that cause runtime errors during the tests (test compilation completes without errors though). After I get the tests running successfully, I'll do a performance comparison.

Philip Hauer's blog post does make several good points — probably the most relevant is his point about the steep learning curve (when compared to maven). The point about IDE support does not apply to gradle 5+ if Kotlin is used for the build scripts instead of groovy: https://docs.gradle.org/5.0/userguide/kotlin_dsl.html#sec:ide_support

**Comment by himanshug** [comment]
yeah agree about steep learning curve .

https://gradle.org/gradle-vs-maven-performance/#performance-advantages-over-maven-that-make-this-possible explains well why reruns of the build would be faster but still not sure why clean build (e.g. as done in travis ) becomes faster with gradle.

**Comment by ccaominh** [comment]
My earlier measurements for maven were unfair as I have not converted most of the maven plugins to their gradle equivalents. The adjusted numbers for maven:
- **Non-parallel:**`mvn install -DskipTests -Dcheckstyle.skip=true -Dforbiddenapis.skip=true -Dpmd.skip=true -Danimal.sniffer.skip=true -Denforcer.skip=true -Dspotbugs.skip=true `(3m 15s): https://scans.gradle.com/s/uetcuf7p57utq
- **Parallel:**`mvn install -DskipTests -Dcheckstyle.skip=true -Dforbiddenapis.skip=true -Dpmd.skip=true -Danimal.sniffer.skip=true -Denforcer.skip=true -Dspotbugs.skip=true -T8` (1m 51s): https://scans.gradle.com/s/nbxz5xiwqtloa

The build for `druid-console` takes 33-34 seconds (mostly front-end related compilation) and needs to be subtracted when comparing against gradle, as I have not translated that yet. After that is taken into account, **gradle 5.5 is 2x faster than maven 3.6.1 for clean builds**.

I was also able to fix the test runtime errors for gradle, but did not measure a significant difference in the test execution time between gradle and maven. **Probably the biggest performance advantage of using gradle over maven, is that gradle is very good at doing incremental builds.**

**Comment by himanshug** [comment]
thanks, so travis builds don't become significantly faster.

for local work , typically I run  specific tests using IDE which run pretty fast or I use maven on specific modules, I am working on, using commands well covered in https://blog.jayway.com/2013/06/09/working-efficiently-with-maven-modules/
that combined with parallel build flag('-Tx') satisfies most of my needs.

at this point, It is indeed true that gradle would be faster on incremental builds in general, but I think devs should list down specific things that they really frequently do(not "could do") while local dev work and see if gradle offers a much better/faster way of doing same, that could help assess whether it makes sense to move to gradle.

in time, maven might catch up on incremental builds too.

**Comment by ccaominh** [comment]
For speeding up travis builds, we can potentially look into seeing which tests are compatible with maven parallel builds. For example, I've seen some tests that bind a fixed port, which would need to be modified or put into a CI job that does not execute test classes in parallel.

**Comment by himanshug** [comment]
that sounds good to me , not sure if related to gradle migration though

**Comment by ccaominh** [comment]
For those interested, my branch that has a gradle build partially implemented: https://github.com/ccaominh/incubator-druid/tree/gradle. Sources and tests compile and unit tests run.

I'm putting this work on hold as the speed gain does not seem to outweigh the cost of having everyone switch from maven. Remaining work includes: refactoring (dependency versions, creating plugins, etc.), integration tests, and converting the various maven plugins that are currently used.
